### PR TITLE
Implement block device abstraction

### PR DESCRIPTION
**Pull request by varqox**
See #1

### CODE REVIEW

File: include/seastar/fs/block_device.hh
Comment (rokinsky) [inline include/seastar/fs/block_device.hh]:
```suggestion
   ~file_block_device_impl() override = default;
```
`virtual` is redundant

File: include/seastar/fs/block_device.hh
Comment (rokinsky) [inline include/seastar/fs/block_device.hh]:
```suggestion
    explicit file_block_device_impl(file f) : _file(std::move(f)) {}
```
Single argument constructors must be marked `explicit`

File: include/seastar/fs/block_device.hh
Comment (rokinsky) [inline include/seastar/fs/block_device.hh]:
```suggestion
    virtual ~block_device_impl() = default
```

File: include/seastar/fs/block_device.hh
Comment (rokinsky) [inline include/seastar/fs/block_device.hh]:
```suggestion
inline future<block_device> open_block_device(sstring name) {
```
Without `inline` keyword here, I have a multiple definition linker error.

File: tests/unit/block_device_test.cc
Comment (psarna) [inline tests/unit/block_device_test.cc]: we have KB, MB and GB defined in `seastar/core/units.hh`, feel free to use them instead of manual shifts:
`min_device_size = 16*MB`

File: tests/unit/block_device_test.cc
Comment (psarna) [inline tests/unit/block_device_test.cc]: ditto, 16*MB looks nicer

File: tests/unit/block_device_test.cc
Comment (psarna) [inline tests/unit/block_device_test.cc]: for a variable that represents a size, use `blocks_num` or `blocks_count` - `blocks` suggests that it's a data structure with blocks inside

File: include/seastar/fs/block_device.hh
Comment (psarna) [inline include/seastar/fs/block_device.hh]: code style request - either drop this change or move it to the previous commit. This commit should only contain changes to the unit test.

**Review by psarna — CHANGES_REQUESTED** [review]
General request - before merging, please squash all fixup commits (like `Fixed block_device unit test`) with `git rebase -i` and repush with `git push -f`. Seastar keeps a clear git history, which means that a single series should not contain fixes to itself - instead, series are always rebased and squashed in order to ensure they are clear.

**Review by rokinsky — CHANGES_REQUESTED** [review]

**Comment by psarna** [comment]
Let's make `block_device` aware of `io_priority_class`, as `seastar::file` is. We'll use the default class for now anyway, but priority classes are crucial for performance, so the API should contain them from the start.

**Comment by varqox** [comment]
How namespaces should be used?
Should block_device and other functionality implemented in zpp_fs be in `seastar` or `seastar::fs` namespace?
Should temporary_file be in `seastar::fs` or `seastar::util` or `seastar::fs::util`?

Regarding pull requests, I understand that commits should be neatly divided so that each adds some part of the whole without touching others (e.g. no fixes) and it should be done before merging. So before the pull is ready for merging, are we supposed to clean the commits or not to clean the commits (to make the review easier) every time we change the pull request? If the second option is better, should one ask for another round of review after cleaning the commits, or just make a merge, or something else?

**Review by rokinsky — COMMENTED** [review]
Looks good.

**Comment by psarna** [comment]
@varqox Let's keep everything in `seastar::fs` so our changes are well separated. Also, no need for `seastar::fs::util`, it's fine for everything to reside in `seastar::fs`.

As for squashing commits, it's actually easier for me to review a clean series instead of a branch full of fixups. It's totally fine to push fixups when developing the change, but it's best if they're cleaned before the final review (the best option would be just to use gerrit, but for now we're restricted to github and its hyper-unintuitive way it presents commits in pull requests).

By the way, for now the list of maintainers which are capable of merging a request to our branch are: me, @haaawk and @kbr-. Once the change is accepted it will be merged by one of us - that way the git tree is more likely to be clean enough to become a merging candidate for seastar/master later.

File: include/seastar/fs/temporary_file.hh
Comment (haaawk) [inline include/seastar/fs/temporary_file.hh]: This uses wrong convention: '&' should be next to temporary_file with no space.

File: include/seastar/fs/temporary_file.hh
Comment (haaawk) [inline include/seastar/fs/temporary_file.hh]: s/temporary_file &/temporary_file&/

File: include/seastar/fs/temporary_file.hh
Comment (haaawk) [inline include/seastar/fs/temporary_file.hh]: remove space between temporary_file and &&

File: include/seastar/fs/temporary_file.hh
Comment (haaawk) [inline include/seastar/fs/temporary_file.hh]: remove space between temporary_file and &&
also remove space between temporary_file and &

**Review by psarna — APPROVED** [review]

**Pull request merged**
